### ISSUE TITLE
Can PSyclone transpose arrays?

### ISSUE THREAD

**Issue by jwallwork23** [stfc/PSyclone#2338]
I have a loop nest which is more suitable for GPU runs if it is reordered so that the inner-most loop gets moved to the outer-most level. This improved the performance but now we no longer have stride-1 access because the array index ordering no longer matches the loop ordering.

I notice that PSyclone can be used to reorder loops. Can it also transpose arrays? If so, the combination of these would save us a lot of effort refactoring our code and would allow use to preserve the existing ordering for CPU runs.

Thanks!

**Comment by sergisiso** [comment]
Hi @jwallwork23 , we don't have this transformation now but we could add it if everything happens locally. This means:
- the array declaration is local (not from an import, at the module level, or from a common block)
- it is never an argument of a call (in which case we don't follow the call-chains - but we could look at inlining if its an easy one), so essentially it must just be used for array math operations (intrinsics can also be tricky as some have a dim parameter that would need to be updated)

**Comment by jwallwork23** [comment]
Thanks for the info, @sergisiso! In our case, the arrays are defined at the module level and used in various places as subroutine arguments.

**Comment by sergisiso** [comment]
Ah, then this is much more tricky to do, since if we miss to update any reference it will break or produce incorrect results.

Are the arrays imported from other modules or just shared among the routines inside the same module? (e.g. if you add the private attribute in the array declaration, does it still compile?)

**Comment by jwallwork23** [comment]
Yeah I can imagine there is a lot more to deal with in this case. We have several files, most of which contain a module with a single subroutine.

**Comment by jwallwork23** [comment]
Good question! I'll investigate further. Sadly in our case we can't collapse many of the loops.
